modcc/printer/cudaprinter.cpp

```diff
@@ -252,4 +362,106 @@ void emit_common_defs(std::ostream& out, const Module& module_) {
     out << popindent << "};\n\n";
 }
 
+void emit_api_body_cu(std::ostream& out, APIMethod* e, bool is_point_proc) {
+    auto body = e->body();
+    auto indexed_vars = indexed_locals(e->scope());
+
+    if (!body->statements().empty()) {
+        out << "int tid_ = threadIdx.x + blockDim.x*blockIdx.x;\n";
+        out << "if (tid_<n_) {\n" << indent;
+        out << "auto gid_ __attribute__((unused)) = params_.node_index_[tid_];\n";
+        out << "auto cid_ __attribute__((unused)) = params_.vec_ci_[gid_];\n";
+
+        auto uses_ion = [&] (ionKind k) {
+            for(auto &symbol : e->scope()->locals()) {
+                if (k == symbol.second->is_local_variable()->ion_channel()) {
+                    return true;
+                }
+            }
+            return false;
+        };
+
+        uses_ion(ionKind::K)  && out << "auto kid_  = params_.ion_k_index_[tid_];\n";
+        uses_ion(ionKind::Ca) && out << "auto caid_ = params_.ion_ca_index_[tid_];\n";
+        uses_ion(ionKind::Na) && out << "auto naid_ = params_.ion_na_index_[tid_];\n";
+
```

Comment:
Rather than hard-coding the index names and ion kinds, which will need to be replaced when we start supporting arbitrary ion species, I'd rather use code analogous to that in `cprinter.cpp:emit_simd_api_body()`, which explicitly gathers the indexed variables and their 'index names' as provided by `decode_indexed_variable`. This also covers the `gid_` and `cid_` case (`cid_` is I believe currently unused, as we're still pull `dt` from the cv-indexed variable), and means we don't have to worry about `__attribute__((unused))` etc.

This also obviates the need for the `uses_ion` lambda.

Reply:
All the code that generates the kernels has been significantly improved while at the same time accomodating the new mechanism infrastructure.
This can be improved in a seperate PR.